### ISSUE TITLE
Better abstraction to deal with persistent service bot channels

### ISSUE THREAD

**Issue by GLolol** [GLolol/PyLink#265]
Summary of issues:
- When a services client is killed, it won't join any relay leaf channels. Perhaps Relay needs to automatically queue channels for services bots when they join a relay channel. 
- ~~Services bots that joined because of Relay aren't removed automatically when that channel is delinked, etc.~~ Actually, the reverse is true with recent 2.0 versions; relay will remove service bots from the home channel on DELINK when it really shouldn't.

Related discussions:

```
22:28 <@GL> i literally don't see a good solution
22:29 <@GL> right now, relay temporarily joins services bots to relay leaf channels, but if the service bot is killed or whatnot it won't rejoin any leaves
22:31 <@GL> but if relay blindly adds / removes channels (based on LINK/DELINK) to the list that the service bot is supposed to join, it might remove channels that the service bot joined by itself
```

Update 2017-01-10: one solution might be to check the `ServiceBot.extra_channels` value on `initializeChannel()`, `removeChannel()`, and on all `PYLINK_SERVICE_JOIN` payloads.

Update 2018-04-04: See https://github.com/GLolol/PyLink/issues/265#issuecomment-378819135 for the proposed 2.0 solution

**Comment by GLolol** [comment]
I may fix this in 2.0 by making service bots leave empty channels and rejoin them when someone else joins (i.e. memorizing "active" channels), sort of like what ChanServ does but minus the takeover protections. 

Possibly related: #146, #218

**Comment by GLolol** [comment]
Implementation wise, this will likely involve:

- [ ] Renaming the `extra_channels` variable in `ServiceBot` to `dynamic_channels`, and properly documenting this as a method for service bots to register "semi-persistent" channels (i.e. they will (re)join when there's people and leave when everyone else does)
- [ ] Adding more complete abstraction to join / part service bots from channels while sending the correct hooks.
- [ ] Expanding the state cleanup hooks in coremods/handlers to make service bots part a channel once it is empty (without removing it from `dynamic_channels`) and rejoin when a join-type event is received.
- [ ] Working on the case of kills: perhaps a similar state cleanup hook is also needed. (Or perhaps we should add an abstract REMOVE_USER hook to simplify this going forward)

----

- [ ] When this is done, Relay's deinit routines can simply be told to ignore service bots completely - if there is no need for the mirrored service bots to remain in a channel, they will leave automatically with the last remaining user and *not* rejoin (since the channel is not in the local network's `dynamic_channels`).
